packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1855,10 +1876,25 @@ export class ContainerRuntime
 		// or something different, like a system message.
 		const runtimeMessage = messageArg.type === MessageType.Operation;
 
-		// Do shallow copy of message, as the processing flow will modify it.
-		const messageCopy = { ...messageArg };
-		const message = this.remoteMessageProcessor.process(messageCopy);
+		// Ungroup the message (we ungroup before sending to remoteMessageProcessor so we can record for summarizer heuristics)
+		for (const ungroupedMessage of this.opGroupingManager.ungroupOp(messageArg)) {
+			if (this._summarizer !== undefined && this.groupedBatchingEnabled) {
```

Comment:
Can we avoid this dependency here? runtime emits "op" event for every op, can we leverage that instead in summarizer?
If it's not sufficient, can we add new type of event to enable this scenario?